### PR TITLE
Lazily download artifacts (Part II): The Clone Wars

### LINKED ISSUE #6491: Would be nice if `flutter run` only downloaded the necessary artifacts
I was on a train with bad wifi.  It was painful.

I was trying to run the gallery on an iOS Simulator.

```
%flutter run
Building flutter tool...
Downloading package sky_engine...                    8101ms
Running 'pub get' in sky_engine...                   450ms
Downloading package sky_services...                  596ms
Running 'pub get' in sky_services...                 3139ms
Downloading package flutter_services...              1929ms
Running 'pub get' in flutter_services...             2294ms
Building Dart SDK summary...                         1753ms
Downloading engine artifacts android-arm...          177027ms
Downloading engine artifacts android-arm-profile...  164117ms
Downloading engine artifacts android-arm-release...  115925ms
Downloading engine artifacts android-x64...          78964ms
Downloading engine artifacts android-x86...          78600ms
Downloading engine artifacts ios...                  178979ms
Downloading engine artifacts ios-profile...          189778ms
Downloading engine artifacts ios-release...          152586ms
Downloading darwin-x64 tools...                      232022ms
Downloading android-arm-profile/darwin-x64 tools...  34656ms
Downloading android-arm-release/darwin-x64 tools...  139862ms
Running 'pub get' in flutter_gallery...              870ms
```

FYI @danrubel

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/6491

Only download the minimum*** set of artifacts needed for a particular host, target, and build mode.

*** I still include gradle and fonts in the minimum set.

-----

Adds additional arguments to several Cache methods:
- buildMode & targetPlatform for filtering by artifacts, can be null. If null, retrieve all artifacts matching one of the other parameters. For example, targetPlatform.ios and null buildMode downloads debug, profile, and release for iOS.
-skipUnknown, which allows downloading artifacts when targetPlatform and hostPlatform are unknown.
- clobber, whether to force re-downloading of artifacts

`flutter precache` will still download the full set of artifacts for a given host platform.
`flutter precache -f/--force` will re download all artifacts.

Re-organizes engine artifacts into a class from lists of lists of strings.

Flutter commands now check the artifacts for the given or default build mode, and any connected devices. (note: android-arm-profile is downloaded unconditionally because we currently need it for a Flutter doctor check).

This is accomplished via a  new method on FlutterCommand called `updateCache`. This is only called if `shouldUpdateCache` is true. The default implementation is:

```
  Future<void> updateCache() async {
    // Only download the minimum set of binaries.
    await cache.updateAll(
      clobber: false,
      skipUnknown: true,
    );
  }
```
Run, for example, looks at the current set of attached devices and ensures the binaries are downloaded for each kind.

Example doctor command:
```
jonahwilliams@jonahwilliams0:~/Documents/flutter/examples/hello_world$ rm -rf ../../bin/cache/
jonahwilliams@jonahwilliams0:~/Documents/flutter/examples/hello_world$ flutter doctor
Downloading Dart SDK from Flutter engine 15f2b92cce916982b7dd8ce658bbf2a465c06ba4...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 50.4M  100 50.4M    0     0  43.4M      0  0:00:01  0:00:01 --:--:-- 43.4M
Building flutter tool...
Downloading Material fonts...                                       1.7s
Downloading package sky_engine...                                   0.3s
Downloading common tools...                                         2.0s
Downloading linux-x64 tools...                                      1.0s
Downloading android-arm-profile/linux-x64 tools...                  1.7s
Downloading Gradle Wrapper...                                       1.0s
Doctor summary (to see all details, run flutter doctor -v):
```

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Includes fix to `flutter drive` not fetching all artifacts

File: packages/flutter_tools/lib/src/cache.dart [hunk 1]
```diff
@@ -210,13 +224,26 @@ class Cache {
...
+    if (!_lockEnabled) {
       return;
+    }
     try {
       for (CachedArtifact artifact in _artifacts) {
-        if (!artifact.isUpToDate())
-          await artifact.update();
+        bool localClobber = clobber;
+        if (localClobber) {
+          await artifact.update(buildModes: buildModes, targetPlatforms: targetPlatforms, skipUnknown: skipUnknown, clobber: localClobber);
```
Comment (gspencergoog) [inline packages/flutter_tools/lib/src/cache.dart]: I know we're pretty lax about line length, but these `update` calls are getting a little long to read.

File: packages/flutter_tools/lib/src/cache.dart [hunk 2]
```diff
@@ -351,110 +416,399 @@ class MaterialFonts extends CachedArtifact {
...
-  List<List<String>> get _dartSdks => <List<String>> [
-    <String>['darwin-x64', 'dart-sdk-darwin-x64.zip'],
-    <String>['linux-x64', 'dart-sdk-linux-x64.zip'],
-    <String>['windows-x64', 'dart-sdk-windows-x64.zip'],
+  /// A set of all possible artifacts to download.
+  ///
+  /// Adding a new artifact:
+  ///
+  /// To ensure that we do not waste a user's time/data/storage, the flutter
+  /// tool should only binaries when they are required. These can be requested
```
Comment (gspencergoog) [inline packages/flutter_tools/lib/src/cache.dart]:
```suggestion
  /// tool should only download binaries when they are required. These can be requested
```

File: packages/flutter_tools/lib/src/cache.dart [hunk 3]
```diff
@@ -351,110 +416,399 @@ class MaterialFonts extends CachedArtifact {
...
+  ///
+  ///   * `hostPlatform` should be one of `TargetPlatform.linux_x64`,
+  ///   `TargetPlatform.darwin_x64`, or `TargetPlatfrom.windows_x64`. In the
+  ///   case where there is no restriction it can be left as null.
+  ///   * `buildMode` should be one of `BuildMode.debug`, `BuildMode.profile`,
+  ///   `BuildMode.release`, `BuildMode.dynamicRelease`, or
+  ///   `BuildMode.dynamicProfile`. In the case where it is required regardless
+  ///   of buildMode, it can be left null.
+  ///   * `targetPlatform` should be one of the supported target platforms.
+  ///   * If, despite the restrictions above the artifact should still be
```
Comment (gspencergoog) [inline packages/flutter_tools/lib/src/cache.dart]:
```suggestion
  ///   * If, despite the restrictions above, the artifact should still be
```

**Review by gspencergoog — APPROVED** [review]

**Comment by gspencergoog** [comment]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

Modulo doc nits above.

File: packages/flutter_tools/lib/src/cache.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/cache.dart]: Since the tools don't contain widget trees, it seems like we should really consider dartfmt

**Pull request merged**